go/vt/vtgateproxy/discovery.go

```diff
@@ -324,32 +318,48 @@ func (b *JSONGateResolverBuilder) GetTargets(poolType string) []targetHost {
 	targets = append(targets, b.targets[poolType]...)
 	b.mu.RUnlock()
 
-	// Shuffle to ensure every host has a different order to iterate through, putting
-	// the affinity matching (e.g. same az) hosts at the front and the non-matching ones
-	// at the end.
-	//
-	// Only need to do n-1 swaps since the last host is always in the right place.
+	b.sorter.shuffleSort(targets, b.affinityField, b.affinityValue)
+
+	return targets
+}
+
+type shuffleSorter struct {
+	rand *rand.Rand
+	mu   *sync.Mutex
+}
+
+func newShuffleSorter() *shuffleSorter {
+	return &shuffleSorter{
+		rand: rand.New(rand.NewSource(time.Now().UnixNano())),
+		mu:   &sync.Mutex{},
+	}
+}
+
+// shuffleSort shuffles a slice of targetHost to ensure every host has a
+// different order to iterate through, putting the affinity matching (e.g. same
+// az) hosts at the front and the non-matching ones at the end.
+func (s *shuffleSorter) shuffleSort(targets []targetHost, affinityField, affinityValue string) {
 	n := len(targets)
 	head := 0
+	// Only need to do n-1 swaps since the last host is always in the right place.
 	tail := n - 1
 	for i := 0; i < n-1; i++ {
-		j := head + b.rand.Intn(tail-head+1)
+		s.mu.Lock()
+		j := head + s.rand.Intn(tail-head+1)
+		s.mu.Unlock()
```

Comment:
This is likely irrelevant since we're not ever calling this under load, but my .02 would be that we don't need to take and drop the lock so often, and instead just taking it once at the beginning of sort and dropping at end would be fine.